### ISSUE TITLE
Panic in some conditions

### ISSUE THREAD

**Issue by ajoga** [daveshanley/vacuum#356]
Hello,

This is an MRC of a situation where vacuum panics.

```shell
$ docker run --rm -v $PWD:/work:ro dshanley/vacuum lint specs/openapi-spec.yml  --details --base=specs/ -r spectral-aws-apigateway-ruleset/aws_important_notes.yml 


██╗   ██╗ █████╗  ██████╗██╗   ██╗██╗   ██╗███╗   ███╗
██║   ██║██╔══██╗██╔════╝██║   ██║██║   ██║████╗ ████║
██║   ██║███████║██║     ██║   ██║██║   ██║██╔████╔██║
╚██╗ ██╔╝██╔══██║██║     ██║   ██║██║   ██║██║╚██╔╝██║
 ╚████╔╝ ██║  ██║╚██████╗╚██████╔╝╚██████╔╝██║ ╚═╝ ██║
  ╚═══╝  ╚═╝  ╚═╝ ╚═════╝ ╚═════╝  ╚═════╝ ╚═╝     ╚═╝


version: latest | compiled: Tue, 31 Oct 2023 16:11:53 UTC
🔗 https://quobix.com/vacuum | https://github.com/daveshanley/vacuum


 INFO  Linting against 43 rules: https://quobix.com/vacuum/rulesets/recommended
panic: runtime error: index out of range [1] with length 1

goroutine 126 [running]:
github.com/pb33f/libopenapi/utils.FindKeyNode({0x0, 0x0}, {0xc000134258, 0x1, 0x0?})
        /go/pkg/mod/github.com/pb33f/libopenapi@v0.12.1/utils/utils.go:255 +0xde
github.com/daveshanley/vacuum/functions/core.Undefined.RunRule({}, {0xc0001a01e0, 0x1, 0x1?}, {0xc0001974a0, 0xc000157810, {0x1533720, 0xc00006bb90}, {0x0, 0x0}, ...})
        /opt/vacuum/functions/core/undefined.go:40 +0x116
github.com/daveshanley/vacuum/motor.buildResults({0xc000157810, 0xc00048f220, {0x192ccb0, 0xc000119170}, 0xc000013098, 0xc00003d620, 0xc000013578, 0xc0002e8a00, 0xc0001578c0, 0x0, ...}, ...)
        /opt/vacuum/motor/rule_applicator.go:413 +0x64b
github.com/daveshanley/vacuum/motor.runRule({0xc000157810, 0xc00048f220, {0x192ccb0, 0xc000119170}, 0xc000013098, 0xc00003d620, 0xc000013578, 0xc0002e8a00, 0xc0001578c0, 0x0, ...})
        /opt/vacuum/motor/rule_applicator.go:338 +0x3b8
created by github.com/daveshanley/vacuum/motor.ApplyRulesToRuleSet
        /opt/vacuum/motor/rule_applicator.go:267 +0xdcf
```

openapi-specs.yml (non-public specification I'm working on, very much trimmed for this ticket):

```yaml
openapi: 3.0.3
info:
  title: sdfghyt
  version: 1.0.0
paths:
  /operations:
    get:
      operationId: listOperations
      summary: Get the list of operations
      responses:
        "200":
          description: OK,
          content:
            application/json:
              schema:
                type: array
              example:
                - uuid: 00000000-0000-0000-0000-000000000000
                  stat: queued
```

aws_important_notes.yml (trimmed down version of https://github.com/andylockran/spectral-aws-apigateway-ruleset/blob/1a5c1cd1580cf708b43426465073ec7f18bce570/aws_important_notes.yml#L95):

```yaml
extends:
  - spectral:oas
rules:
  aws-example-tag:
    description: "The example tag is not supported by API gateway, but there is no harm in including it."
    severity: hint
    given: $..example
    then:
      function: undefined
```

It seems to me that both the rule and the API spec are relatively legal, even trimmed like this, and should not make vacuum panic.
I have the very same panic on the whole file, even though vacuum runs very well on it with the recommended built-in rules.

**Comment by daveshanley** [comment]
Thank you, will be fixed.

**Comment by daveshanley** [comment]
I have looked into this, and it's getting upset because it cannot parse the yaml. With the latest PR #357, when parsing the doc, I get this error: 

![Screenshot 2023-11-04 at 5 38 42 PM](https://github.com/daveshanley/vacuum/assets/187345/c96c20c5-466e-4bff-bc2d-9ec4b5018681)

 `linting error: unable to parse specification: yaml: line 26: found unknown escape character`

which turns out to be this: 

`"^\/[A-Za-z0-9.,{}\/]*"`

If it's changed to: `"^\\/[A-Za-z0-9.,{}\\/]*"` it works. This is not something I am in control of, this is the https://pkg.go.dev/gopkg.in/yaml.v3 parser complaining

**Comment by ajoga** [comment]
Yes, the full ruleset makes vacuum chokes in various places, but I filled this issue more specifically for the « aws-example-tag » rule.If you only use that rule, you’ll be able to reproduce the panic.I kind of figured the other issues were more subtle and didn’t think they were worth a report hereLe 4 nov. 2023 à 22:43, quobix ***@***.***> a écrit :﻿
I have looked into this, and it's getting upset because it cannot parse the yaml. With the latest PR #357, when parsing the doc, I get this error:

linting error: unable to parse specification: yaml: line 26: found unknown escape character
which turns out to be this:
"^\/[A-Za-z0-9.,{}\/]*"
If it's changed to: "^\\/[A-Za-z0-9.,{}\\/]*" it works. This is not something I am in control of, this is the https://pkg.go.dev/gopkg.in/yaml.v3 parser complaining

**Comment by daveshanley** [comment]
OK, thank you I will try again

**Comment by daveshanley** [comment]
I found the issue, thank you, it's a tiny glitch in `libopenapi` will be fixed shortly.

**Comment by daveshanley** [comment]
Not sure why this auto-closed, anyway new version of vacuum inbound with this fix coming.

**Comment by daveshanley** [comment]
This has been fixed in vacuum v`0.4.3` Thank you for reporting it! I also updated the undefined function to output a more coherent message based on your example.

**Comment by ajoga** [comment]
Indeed it's fixed, thank you @daveshanley !
